### DISCUSSION TITLE
Cannot direcly return asyncpg.Record in endpoints

### DISCUSSION THREAD

**Discussion by MichaelHhHhH** [fastapi/fastapi#10301] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class Item(pydantic.BaseModel):
    id: int


@app.get('/{item_id}', response_model=Item)
async def get_item(item_id: int):
    record = await get_item_from_db(item_id=item_id)  # asyncpg.Record
    return record
```

### Description

Cannot directly use asyncpg.Record as result of endpoint with response_model as pydantic.BaseModel

```
class Item(pydantic.BaseModel):
    id: int


@app.get('/{item_id}', response_model=Item)
async def get_item(item_id: int):
    record = await get_item_from_db(item_id=item_id)  # asyncpg.Record
    return record
```

Error

```
fastapi.exceptions.ResponseValidationError: 3 validation errors:
E                 {'type': 'missing', 'loc': ('response', 'id'), 'msg': 'Field required', 'input': <Record id=1>, 'url': 'https://errors.pydantic.dev/2.3/v/missing'}
```

### Operating System

Linux

### Operating System Details

Fedora 35

### FastAPI Version

0.103.1

### Pydantic Version

2.3.0

### Python Version

3.11.1

### Additional Context

Help please

**Comment by alex-pobeditel-2004** [comment]
Hello!

I think it is a good practice to transform your `Record._mapping` into a correct Pydantic model before returning response - other way FastAPI has no clue which types your record has and it can't generate proper OpenAPI schema for the endpoint.

**Reply by MichaelHhHhH** [reply]
Hello, thank you for answer, I meant that in pydantic 1 it worked correctly and without any problems. So it is broken only if I use pydantic 2. Main question was about, how to add something like type adapter between result of endpoint and pydantic model validation. So that I can return asyncpg.Record directly in endpoints. But, I think it's impossible now. Now, can I also declare response_model as pydantic.BaseModel but without validation? Only for openapi schema?

**Comment by nemosanima** [comment]
Hi!

The error you are encountering is because FastAPI's response_model expects the returned value to be an instance of the specified Pydantic model (Item in this case). However, you are trying to return an asyncpg.Record object directly, which is not a Pydantic model and doesn't match the expected schema.

To resolve this issue, you should convert the asyncpg.Record object to an instance of the Item Pydantic model before returning it from your endpoint. You can do this by creating an instance of the Item model and populating its fields with the values from the asyncpg.Record. Here's how you can modify your code to achieve this: 

```python
@app.get('/{item_id}', response_model=Item)
async def get_item(item_id: int):
    record = await get_item_from_db(item_id=item_id)
    
    # Convert the asyncpg.Record to an instance of the Item model
    item = Item(id=record['id'])
    
    return item
```

**Reply by MichaelHhHhH** [reply]
Hello, thank you for answer, I meant that in pydantic 1 it worked correctly and without any problems. So it is broken only if I use pydantic 2. Main question was about, how to add something like type adapter between result of endpoint and pydantic model validation. So that I can return asyncpg.Record directly in endpoints. But, I think it's impossible now. Now, can I also declare response_model as pydantic.BaseModel but without validation? Only for openapi schema?

**Comment by MichaelHhHhH** [comment]
Hello, thank you for answer, I meant that in pydantic 1 it worked correctly and without any problems. So it is broken only if I use pydantic 2. Main question was about, how to add something like type adapter between result of endpoint and pydantic model validation. So that I can return asyncpg.Record directly in endpoints. But, I think it's impossible now. Now, can I also declare response_model as pydantic.BaseModel but without validation? Only for openapi schema?

**Comment by nemosanima** [comment]
You can declare just for schema with Any typing:

```python
from typing import Any


class Item(BaseModel):
    id: int
    title: str


@app.get('/{item_id}', response_model= Item | Any)
def get_item(item_id: int):
    return 'hello'
```

or without response_model:

```python
from typing import Any


@app.get('/{item_id}')
def get_item(item_id: int) -> Item | Any:
    return 'hello'
```

swagger:
<img width="1433" alt="swagger" src="https://github.com/tiangolo/fastapi/assets/115027077/3f047891-7d31-4cff-982f-e5ae0870d406">

**Comment by MichaelHhHhH** [comment, marked as answer]
Hello, this solution doesn't work. I finally found how can I do this. Need to create your own class Record that inherits from asyncpg.Record and implement method __getattr__
